---
title: "Microbiome composition"
bibliography: 
- bibliography.bib
- references.bib
output: 
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---
  <!--
  %\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{microbiome tutorial - composition}
%\usepackage[utf8]{inputenc}
%\VignetteEncoding{UTF-8}  
-->
  
  
## Microbiota composition
  

Also see [phyloseq barplot examples](http://joey711.github.io/phyloseq/plot_bar-examples.html).
  
Read example data from a [diet swap study](http://dx.doi.org/10.1038/ncomms7342):
  
```{r composition-example1}
# Example data
library(microbiome)
library(dplyr)
data(dietswap)
pseq <- dietswap

# Pick sample subset
library(phyloseq)
pseq2 <- subset_samples(pseq, group == "DI" & nationality == "AFR")
```

### Barplots for composition

Show OTU absolute abundance in each sample. Plot absolute taxon
abundances:
  
```{r composition-example1b, fig.width=16, fig.height=8, message=FALSE}
library(ggplot2)
theme_set(theme_bw(22)) # Black/white color theme
p <- plot_composition(pseq2, taxonomic.level = "Phylum") +
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(nrow = 10, byrow = TRUE))
print(p)       
```

Arrange by sample variable. Focus on the core taxa. Africans have more Prevotella as expected. Absolute counts:
  
```{r composition-example4, fig.width=13, fig.height=5}
# Limit the analysis on core taxa and specific sample group
pseq2 <- pseq %>%
  core(detection = 10^4, prevalence = .5) %>%
  subset_samples(group == "DI" & timepoint.within.group == 1)

p <- plot_composition(pseq2,
                      sample.sort = "nationality", # Sort by nationality
                      x.label = "nationality") +   # Label by nationality
  guides(fill = guide_legend(nrow = 5, byrow = TRUE)) +
  theme(legend.position = "bottom")
print(p)
```


Same with compositional (relative) abundances:
  
```{r composition-example4b, fig.width=12, fig.height=5, out.width="300px", fig.show="hold"}
p <- plot_composition(pseq2,
                      sample.sort = "nationality",
                      x.label = "nationality",
                      transform = "compositional") +
     guides(fill = guide_legend(ncol = 1))
print(p)

# Or try another theme
# from https://github.com/hrbrmstr/hrbrthemes
library(hrbrthemes)
library(gcookbook)
library(tidyverse)
p2 <- p + scale_y_percent() +
          labs(x="Samples", y="Relative abundance (%)",
                                   title="Relative abundance data",
                                   subtitle="Subtitle",
                                   caption="Caption here 'g'") + 
  theme_ipsum(grid="Y")
print(p2)  
```


Averaged by group:
  
```{r composition-example4c, fig.width=12, fig.height=5}
p <- plot_composition(pseq2,
                      average_by = "nationality", transform = "compositional")
print(p)
```



### Composition heatmaps


Plain heatmap on absolute abundances

```{r composition-example5, fig.width=10, fig.height=3}
theme_set(theme_bw(30))
p <- plot_composition(pseq2, plot.type = "heatmap", mar = c(6, 13, 1, 1))
```


Heatmap with clr-transformed OTUs

```{r composition-example6, fig.width=10, fig.height=3}
p <- plot_composition(pseq2, plot.type = "heatmap", transform = "clr", mar = c(6, 13, 1, 1))
```


Same with relative abundance, samples and OTUs sorted with the neatmap method:
  
```{r composition-example7, fig.width=10, fig.height=3, eval = FALSE}
p <- plot_composition(pseq2, plot.type = "heatmap", transform = "compositional"
                      sample.sort = "neatmap", otu.sort = "neatmap",
                      mar = c(6, 13, 1, 1))
```


Same with Z-transformed, samples and OTUs sorted manually based on compositional data (Z-transformed data has negative values and the sorting method is not yet implemented for that):
  
```{r composition-example8, fig.width=10, fig.height=3, warning=FALSE, eval=FALSE}
pseq3 <- transform(pseq2, "compositional")
sample.sort <- neatsort(pseq3, method = "NMDS", distance = "bray", target = "sites") 
otu.sort <- neatsort(pseq3, method = "NMDS", distance = "bray", target = "species")
p <- plot_composition(pseq2, plot.type = "heatmap", transform = "Z-OTU",
                      sample.sort = sample.sort, otu.sort = otu.sort,
                      mar = c(6, 13, 1, 1))
```

### Plot taxa prevalence

Load [example data](Data.md):  
  For this example, we will use two different datasets.   

For HITChip based 16S rRNA gene profilling data. 
The [diet swap study](http://dx.doi.org/10.1038/ncomms7342):  
  
```{r prevalence-example1, fig.height=6, fig.width=8, warning=FALSE}
# Example data
data(dietswap)
pseq <- dietswap
plot <- plot_taxa_prevalence(pseq, 'Phylum')
print(plot)
```

The Dynamics IBD data from [Halfvarson J., et al. Nature Microbiology, 2017](http://www.nature.com/articles/nmicrobiol20174). It was downloaded from [Qiita](https://qiita.ucsd.edu/study/description/1629). The Qiita Study ID is 1629.  

```{r, fig.height=6, fig.width=8}
data(DynamicsIBD)
p0 <- DynamicsIBD  
p0.f <- format_phyloseq(p0)
# Order/Class/Family level for the DynamicsIBD dataset are available, see
# tax_table(p0.f)
plot <- plot_taxa_prevalence(p0.f, 'Phylum')
print(plot)
```


