## Diversity 

Use probe-level data (pseq.probe) to quantify diversity:

```{r example-5, echo=TRUE, message=FALSE}
# Diversity estimation
diversities <- estimate_diversity(pseq.probe)[, c("Shannon", "InvSimpson")]

# Show diversity estimates for the first 5 samples:
library(knitr)
kable(diversities[1:5,])
```


## Richness

Estimate richness (probe count):

```{r example-6, echo=TRUE, message=FALSE}
# Estimate richness
res <- estimate_diversity(pseq.probe, det.th = 1)
richness <- res$Observed
names(richness) <- rownames(res)

# Show richness estimates for the first 5 samples
richness[1:5]
```


Visualize diversity vs. discrete metadata variable (here 'diet' but you can call any variable from your metadata):

```{r example-7, echo=TRUE, message=FALSE}
plot_diversity(pseq.probe, x = "diet", measures = "Shannon")
plot_diversity(pseq.probe, x = "diet", measures = "InvSimpson")
```

### Wilcoxon test

Wilcoxon test for each OTU based on gender comparison. You can replace the gender with any variable that is available in your sample metadata and has exactly two groups.


Run Wilcoxon test and show the first 5 values (sorted by p-value):

```{r example-8, echo=TRUE, message=FALSE}
res <- check_wilcoxon(pseq.L2, "gender", p.adjust.method = "BH", sort = TRUE)
kable(res[1:5,])
```

[Volcano plot](https://en.wikipedia.org/wiki/Volcano_plot_%28statistics%29) shows how the fold-changes and p-values are related. In general, if there are significant findings, the fold-changes tend to be larger with smaller p-values:

```{r example-9, echo=TRUE, message=FALSE}
library(ggplot2)
ggplot(res, aes(x = p.value, y = fold.change.log10)) + 
  geom_point() + 
  xlim(0, 1)
```

Boxplots of the most significant groups:

```{r example-10, echo=TRUE, message=FALSE}
top.groups <- rownames(res)[1:4]
for (tax in top.groups) {
  p <- boxplot_abundance(pseq.L2, x = "gender", y = tax, log10 = TRUE, title = tax)
  print(p)
}
```


### Hierarchical clustering

Cluster the samples in probe-level data with hierarchical clustering:

```{r example-, echo=TRUE, message=FALSE}
# Calculate dissimilarity matrix
d <- 1 - cor(as.matrix(otu_table(pseq.probe)@.Data), method = "spearman")
# Perform hierarchical clustering
hc <- hclust(as.dist(d), method = "complete")
# Show the plot
plot(hc, hang = -1)
```




Composition with relative abundances:

```{r composition-example3, fig.path = "figure/", fig.width=8, fig.height=10}
p <- plot_composition(pseq, taxonomic.level = "Phylum", relative.abundance = TRUE)
p <- p + guides(fill = guide_legend(nrow = 7, byrow = TRUE))
p <- p + theme(legend.position = "bottom")
print(p)
```

Read HITChip data from the specified folder and probe summarization
method (returns the precalculated data matrices for all levels, sample
metadata and taxonomy; without detection thresholding):

```{r dataread, eval=TRUE, message=FALSE, warning=FALSE}
# Read precalculated HITChip data and check available data entries
chipdata <- read_profiling(method = "frpa", data.dir = data.directory)
print(names(chipdata))

# Pick specific data field (for instance oligo level or L2-level data)
probedata <- chipdata[["probedata"]]
dat <- chipdata[["L2"]]

# Check the output
kable(head(dat))
```



# Read HITChip data in phyloseq format
# Pick specific data field (for instance oligo level or L2-level data)
probe.data <- chipdata[["probedata"]]
l2.data <- chipdata[["L2"]]
pseq <- read_hitchip(data.directory, method = "frpa", detection.threshold = 10^1.8)$pseq

dats <- read_profiling(data.dir = data.path, method = "frpa")$species

s <- intersect(rownames(dat), rownames(otu_table(pseq.L2)))
plot(unlist(log10(dat[s,])), log10(unlist(otu_table(pseq.L2)@.Data[s, ])))

s <- intersect(rownames(dats), rownames(otu_table(pseq)))
plot(unlist(log10(dats[s,])), log10(unlist(otu_table(pseq)@.Data[s, ])))

s <- intersect(rownames(dats), rownames(otu_table(data0$pseq)))
plot(unlist(log10(dats[s,])), log10(unlist(otu_table(data0$pseq)@.Data[s, ])))

data0 <- read_hitchip(data.dir = data.path, method = "frpa", detection.threshold = 0)

dat <- read_profiling(data.dir = data.path, method = "frpa")$L2



```{r bimodality-dip, message=FALSE, warning=FALSE, fig.path = "figure/"}
# DIP multimodality test uses the separate diptest package.
library(diptest)

# Pick OTU log10 data
otu <- otu_table(pseq0)@.Data
dip <- apply(otu, 1, function (x) dip.test(x, simulate.p.value = TRUE, B = 1000))
dip2 <- data.frame(t(sapply(dip, function (x) {c(x$statistic, x$p.value)})))
colnames(dip2) <- c("score", "p.value")
dip2$tax <- names(dip)

# Dip quantifies unimodality. Values range between 0 to 1. 
# Values less than 0.05 indicate significant deviation from unimodality. 
# To score multimodality, use the inverse:
multimodality.dip <- 1 - dip2$score
```


Compare the alternative bimodality/multimodality scores

```{r bimodality-comp, message=FALSE, warning=FALSE}
pairs(cbind(pb = bimodality.pb, bc = bimodality.sarle, dip = bimodality.dip))
```


```{r pca-physeq2, message=FALSE, message=FALSE, fig.height=8, fig.width=8, fig.path = "figure/"}
# Example data
pseq <- download_microbiome("dietswap")

# Just include the top Phyla
# TODO: replace tapply with non-plyr functions
phylum.sum <- tapply(taxa_sums(pseq), tax_table(pseq)[, "Phylum"], sum, na.rm = TRUE)
top5phyla <- names(sort(phylum.sum, TRUE))[1:5]
pseq2 <- prune_taxa((tax_table(pseq)[, "Phylum"] %in% top5phyla), pseq)
pseq2.ord <- prune_taxa((tax_table(pseq.ord)[, "Phylum"] %in% top5phyla), pseq)

p <- plot_ordination(pseq2, pseq.ord2, type = "biplot",
     	     color = "bmi_group", shape = "Phylum", title = "biplot")

# modify the automatic shape scale
pseq.shape.names <- get_taxa_unique(pseq, "Phylum")
pseq.shape <- 15:(15 + length(pseq.shape.names) - 1)
names(pseq.shape) <- pseq.shape.names
pseq.shape["samples"] <- 16
p + scale_shape_manual(values = pseq.shape)
```


```{r pca-physeq4, message=FALSE, message=FALSE, fig.height=8, fig.width=8, fig.path = "figure/"}
p4 <- plot_ordination(pseq, pseq.ord, type = "split",
      			    color = "Phylum", shape = "bmi_group", 
			    label = "sample", title = "split")
p4

gg_color_hue <- function(n) {
    hues = seq(15, 375, length = n + 1)
    hcl(h = hues, l = 65, c = 100)[1:n]
}
color.names <- levels(p4$data$Phylum)
p4cols <- gg_color_hue(length(color.names))
names(p4cols) <- color.names
p4cols["samples"] <- "black"
p4 + scale_color_manual(values = p4cols)

# Supported ordination methods 
# TODO: get rid of plyr::llply
skip <- T
if (!skip)  {
dist <- "bray"
ord_meths = c("DCA", "CCA", "RDA", "DPCoA", "NMDS", "MDS", "PCoA")
plist <- llply(as.list(ord_meths), function(i, physeq, dist) {
    ordi = ordinate(physeq, method = i, distance = dist)
    plot_ordination(physeq, ordi, "samples", color = "bmi_group")
}, physeq, dist)
names(plist) <- ord_meths
pdataframe = ldply(plist, function(x) {
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"
p = ggplot(pdataframe, aes(Axis_1, Axis_2, color = SampleType, shape = human, 
    fill = SampleType))
p = p + geom_point(size = 4) + geom_polygon()
p = p + facet_wrap(~method, scales = "free")
p = p + scale_fill_brewer(type = "qual", palette = "Set1")
p = p + scale_colour_brewer(type = "qual", palette = "Set1")
p
}
```

PCA with ggplot2 - the above example gives a shortcut for the following:

```{r density2, message=FALSE, warning=FALSE, fig.width=15, fig.height=8, fig.path = "figure/"}
# Arrange projected data onto a data frame
m <- sample_data(pseq2)
coms <- intersect(rownames(proj), rownames(m))
df <- as.data.frame(cbind(proj[coms,], m[coms,]))
names(df) <- c("x", "y", colnames(m))

# Construct the figure with ggplot2
library(ggplot2)
theme_set(theme_bw(15))
p <- ggplot(df) 

# Add densities
p <- p + stat_density2d(aes(x = x, y = y, fill=..density..), geom="raster", stat_params = list(contour = F), geom_params = list()) 
p <- p + scale_fill_gradient(low="white", high="black") 

# Add points
p <- p + geom_point(aes(x = x, y = y, color = sample_data(pseq2)$gender), size = 1.5) 

# Add labels
p <- p + xlab("PCA 1") + ylab("PCA 2") + ggtitle("Density plot")
p <- p + scale_colour_discrete(name = "Sex")

# Plot the figure
print(p)
```




<!--
### Long-term follow-up time series (David et al. 2014)

The data set from [David et al. Genome Biology 2014, 15:R89](http://genomebiology.com/2014/15/7/R89):

```{r peerjdata2}
library(microbiome)
data.david2014 <- download_microbiome("david2014")
```
-->

<!--Preprocessing of HITChip data is based on RPA [`r citep(citation("RPA"))`; `r citep("10.1109/TCBB.2009.38")`]-->


```{r diversitywithmetadata, warning=FALSE, message=FALSE, fig.width=9, fig.heigth=6, fig.path = "figure/"}
# Add diversity into sample data
sample_data(d)$diversity <- estimate_richness(d, measures = c("Shannon"))$Shannon

# Remove samples with no BMI information
dsub <- subset_samples(d, !is.na(bmi_group))

# Visualize
df <- harmonize_fields(sample_data(dsub))
p <- ggplot(df)
p <- p + geom_boxplot(aes(x = bmi_group, y = diversity))
p <- p + ggtitle("Microbiota diversity vs. obesity")
print(p)
```


### Compare with known background factors

Diversity vs. obesity group (factor):

```{r diversitywithmetadata, warning=FALSE, message=FALSE, fig.width=8, fig.heigth=6, fig.path = "figure/"}
p <- plot_diversity(pseq, x = "bmi_group", measures = c("Shannon", "Simpson"))
print(p)
```



## Hierarchical clustering: reproduce profiling script plot 

```{r clustering-example, message=FALSE, warning=FALSE, fig.width=8, fig.heigth=15, fig.path = "figure/"}
# Download example data
library(microbiome)
pseq <- download_microbiome("peerj32")$physeq

# Pick OTU log10 data
otu <- otu_table(pseq)@.Data
otu.log10 <- log10(otu)

# Hierarchical clustering plot (write into png file)
metric <- "pearson" # note: use log10 with Pearson!
hc <- calculate.hclust(otu.log10, method = "complete", metric = metric)
plot(hc, hang=-1, main = paste("Hierarchical clustering (", metric, ")", sep = ""), las = 1, xlab = "")
```



### Evenness

```{r ev-example, warning=FALSE, message=FALSE, fig.path = "figure/"}
# Pielou's evenness is S/ln(R) w.r.t. given detection threshold
# NOTE: here we use detection threshold for diversity as well because
# the exact same data has to be used for diversity and richness calculation,
# and for richness calculation the detection threshold needs to be set anyway
# Diversity can be as such calculated also without threshold (see above)
# but that gives somewhat different result.
otu2 <- otu - det.th # NOTE: absolute (not log) scale data
S <- vegan::diversity(otu2, index = "shannon")
R <- rowSums(otu2 > 0)
ev <- S/log(R)
```
### Density plot

  # note that taking relative abundances likely changes the abundance histograms
  tax <- "Bacilli"
  theme_set(theme_bw(20))
p <- ggplot(filter(dfm, Taxon == "Vibrio"), aes(x = 100*Abundance))
p <- p + geom_density(fill = "darkgray")
p <- p + scale_x_log10()
p <- p + xlab("Relative Abundance (%)")
p <- p + ggtitle(tax)
print(p)

### Bagged RDA

Calculate Bagged RDA and visualize the results:

```{r rda5, warning=FALSE, message=FALSE, fig.width=8, fig.height=8, eval = FALSE}
# Modify the group vector into a facor as required by Bagged RDA
y <- factor(annot$time); names(y) <- rownames(annot)

# Bagged RDA
Bag.res <- Bagged.RDA.Feature.Selection(t(l2), y, sig.thresh=0.05, nboot=100)

# Visualize
PlotBaggedRDA(Bag.res, y)
```


### Oligo heatmap

This reproduces the oligo-level heatmap from profiling script. If there are problems, try to tune ppcm, figureratio and fontsize (see help(add.heatmap) for details)

```{r oligoheatmap, message=FALSE, warning=FALSE}
library(microbiome)

# Load Phylogeny
phylogeny.info <- GetPhylogeny("HITChip")

# Load oligo-level data

# Replace data.directory here with your own profiling script output data directory
data.directory <- system.file("extdata", package = "microbiome")

oligodata <- read.profiling(level = "oligo", log10 = FALSE, data.dir = data.directory)

# Produce the plot and save it to the working directory
library(HITChipDB)
hc.params <- add.heatmap(log10(oligodata), output.dir = ".", phylogeny.info = phylogeny.info)
```

### A longer version with source code

The following example picks 20 random species and visualizes the HITChip signal as ordered barplot. In this example the bars are colored according to the L2 group. 

```{r barplot-example2, warning=FALSE, message=FALSE, fig.width=8, fig.height=8, fig.path = "figure/"}
# Load Phylogeny
phylogeny.info <- GetPhylogeny("HITChip")

# Get example data 
data(peerj32)
x <- log10(t(peerj32$microbes))

# Pick 20 species from first sample at random
taxa <- rownames(x)[sample(nrow(x), 20)]

# Signal of the selected taxa at first sample
signal <- x[taxa,1]

# Higher-level taxonomic groups for the taxa
l1 <- droplevels(levelmap(taxa, level.from = "L2", level.to = "L1", phylogeny.info = phylogeny.info))

# Collect all into a data.frame
df <- list()
df$taxa <- taxa
df$L1 <- l1
df$signal <- signal
df <- data.frame(df)

# Define colors for groups
l1.colors <- rainbow(length(unique(df$L1)))
names(l1.colors) <- as.character(unique(df$L1))

# Rearrange the data.frame
m <- melt(df)

# Sort by signal (ie. change order of factors for plot)
df <- within(df, taxa <- factor(taxa, levels = taxa[order(abs(signal))]))

# Plot the image
p <- ggplot(aes(x = taxa, y = signal, fill = L1), data = df) 
p <- p + scale_fill_manual(values = l1.colors[as.character(levels(df$L1))])

p <- p + geom_bar(position="identity", stat = "identity") + theme_bw() + coord_flip()
p <- p + ylab("Fold change") + xlab("") + ggtitle("My barplot")
p <- p + theme(legend.position="right")
p <- p + theme(panel.border=element_rect())

print(p)
```
