## Preprocessing taxonomic profiling data

Here we show how to manipulate microbiome data sets using tools from
the [phyloseq package](http://joey711.github.io/phyloseq/), including
subsetting, aggregating and filtering.


### Picking data from phyloseq  

Assuming your data ('pseq' below) is in the phyloseq format, many
standard tools are available.


Sample metadata:

```{r data21}
library(phyloseq)
meta <- sample_data(pseq)
```

Taxonomy table

```{r data22}
tax.table <- tax_table(pseq)
```

Pick taxa abundance data matrix. In this example the OTU level corresponds to genus-like groups (the function name otu_table is somewhat misleading):

```{r data23}
otu <- otu_table(pseq)@.Data
```

Aggregate the abundance matrix to higher-level taxa on HITChip:

```{r data24}
pseq2 <- aggregate_taxa(pseq, "Phylum") # Aggregate into phyloseq object
dat <- otu_table(pseq2)@.Data # Pick aggregated abundance table
```

Melted data for plotting is readily obtained with the phyloseq psmelt function:

```{r data25}
df <- psmelt(pseq)
kable(head(df))
```


### Standard data processing operations


Let us use the example data from [O'Keefe et al. Nat. Comm. 6:6342, 2015](http://dx.doi.org/10.1038/ncomms7342) from [Data Dryad](http://dx.doi.org/10.5061/dryad.1mn1n). This is a two-week diet swap study between western (USA) and traditional (rural Africa) diets including microbiota profiling:

```{r phyloseqdata, fig.width=10, fig.height=7, warning=FALSE, message=FALSE}
library(microbiome)
data("dietswap")
pseq <- dietswap
```

### Sample operations

Sample names and variables

```{r preprocess2b, warning=FALSE, message=FALSE}
head(sample_names(pseq))
```

Sample sums

```{r preprocess2c, warning=FALSE, message=FALSE}
head(sample_sums(pseq))
```

Abundance of a given species in each sample

```{r preprocess2d, warning=FALSE, message=FALSE}
head(get_sample(pseq, taxa_names(pseq)[1]))
```

Filter samples

```{r preprocess3, warning=FALSE, message=FALSE}
f1 <- filterfun_sample(topp(0.1))
taxa <- genefilter_sample(pseq, f1, A = round(0.5 * nsamples(pseq)))
```

Select samples by specific metadata fields:

```{r data-subsetting}
pseq.subset <- subset_samples(pseq, nationality == "AFR")
```


### Data transformations

The microbiome package provides a wrapper for standard sample/OTU transformations. For arbitrary transformations, use the transform_sample_counts function in the phyloseq package.

Log10 transformation (log(1+x) if the data contains zeroes)

```{r preprocessing-logtrans}
pseq.log <- transform_phyloseq(pseq, "log10")
```

Z transformation:

```{r preprocessing-ztrans}
# Z-transform OTUs
pseq.zotu <- transform_phyloseq(pseq, "Z", "OTU")
```

Relative abundances (the input data needs to be in absolute scale, not logarithmic!):

```{r data10}
pseq1 <- transform_phyloseq(pseq, "relative.abundance", "OTU")
pseq2 <- transform_sample_counts(pseq, function(x) x/sum(x))
```


### Variable operations

Sample variable names

```{r preprocess4, warning=FALSE, message=FALSE}
sample_variables(pseq)
```

Pick variable values for a given variable

```{r preprocess4b, warning=FALSE, message=FALSE}
head(get_variable(pseq, sample_variables(pseq)[1]))
# Assign fields to sample metadata
# sample_data(GP)$human <- ..
```

### Taxa operations


Number of taxa

```{r preprocess3e, warning=FALSE, message=FALSE}
ntaxa(pseq)
```


Names

```{r preprocess3f, warning=FALSE, message=FALSE}
ranks <- rank_names(pseq)
taxa <- taxa_names(pseq)
```


Prune taxa:

```{r preprocess3b, warning=FALSE, message=FALSE}
taxa <- levelmap(NULL, "Phylum", "Genus", pseq)$Bacteroidetes

# With given taxon names
ex2 <- prune_taxa(taxa, pseq)

# Taxa with positive sum across samples
ex3 <- prune_taxa(taxa_sums(pseq) > 0, pseq)
```


Subset taxa:

```{r preprocess3c, warning=FALSE, message=FALSE}
pseq <- subset_taxa(pseq, Phylum == "Bacteroidetes")
```


Filter by user-specified function values (here variance):

```{r preprocess3d, warning=FALSE, message=FALSE}
f <- filter_taxa(pseq, function(x) var(x) > 1e-05, TRUE)
```


List unique phylum-level groups: 

```{r preprocess3g, warning=FALSE, message=FALSE}
head(get_taxa_unique(pseq, "Phylum"))
```

Pick detected taxa by sample name:

```{r preprocess3gg, warning=FALSE, message=FALSE}
samplename <- sample_names(pseq)[[1]]
tax.abundances <- get_taxa(pseq, samplename)
```


Taxa sums

```{r preprocess3h, warning=FALSE, message=FALSE}
head(taxa_sums(pseq))
```


### Merging operations

Aggregate OTUs to higher taxonomic levels, use (on HITChip we use L1/L2 instead of Phylum/Genus). See also [merge_samples and and merge_taxa](http://joey711.github.io/phyloseq/merge.html):

```{r preprocessing-aggregate-taxa, eval=FALSE}
pseq.L2 <- aggregate_taxa(pseq, level = "Genus")
```

Merging phyloseq objects

```{r preprocess-merge, warning=FALSE, message=FALSE, eval=FALSE}
merge_phyloseq(pseq1, pseq2)
```


### Rarification

```{r preprocess-rarify, warning=FALSE, message=FALSE}
pseq.rarified <- rarefy_even_depth(pseq)
```


### HITChip taxonomy

Check the overall HITChip taxonomy:

```{r phylogeny-example2, warning=FALSE, message=FALSE}
require(microbiome)
data("hitchip.taxonomy")
tax.table <- hitchip.taxonomy$full
kable(head(tax.table))
```

Conversion between taxonomic levels:

```{r phylogeny-example2b, warning=FALSE, message=FALSE}
m <- levelmap(c("Akkermansia", "Bacteroides fragilis et rel."), 
              from = "L2", to = "L1", tax.table)

# Another example
data(GlobalPatterns)
taxtable <- tax_table(GlobalPatterns)
levelmap("Crenarchaeota", 'Phylum', 'Kingdom', taxtable)
```


