---
title: "Preprocessing"
author: "Leo Lahti"
date: "`r Sys.Date()`"
bibliography: 
- bibliography.bib
- references.bib
output: 
  rmarkdown::html_vignette
---
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial - Preprocessing}
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}  
-->


## Preprocessing taxonomic profiling data

Here we show how to manipulate microbiome data sets using tools from
the [phyloseq package](http://joey711.github.io/phyloseq/), including
subsetting, aggregating and filtering.


### Picking data from phyloseq  

Assuming your data ('pseq' below) is in the phyloseq format, many
standard tools are available.


Sample metadata:

```{r data21}
library(phyloseq)
library(microbiome)
data("atlas1006") 
pseq <- atlas1006
meta <- sample_data(pseq)
```

Taxonomy table:

```{r data22}
tax.table <- tax_table(pseq)
```


Abundances for taxonomic groups ('OTU table') as a taxa x samples
matrix:

```{r data23}
otu <- taxa_abundances(pseq)
```


Melted data for plotting:

```{r data25}
df <- psmelt(pseq)
kable(head(df))
```


### Standard data processing operations

Let us look at [a two-week diet swap
study](http://dx.doi.org/10.1038/ncomms7342) between western (USA) and
traditional (rural Africa) diets including microbiota profiling:

```{r phyloseqdata, fig.width=10, fig.height=7, warning=FALSE, message=FALSE}
library(microbiome)
data("dietswap")
pseq <- dietswap
```

### Sample operations

Sample names and variables

```{r preprocess2b, warning=FALSE, message=FALSE}
head(sample_names(pseq))
```

Sample sums

```{r preprocess2c, warning=FALSE, message=FALSE}
head(sample_sums(pseq))
```

Abundance of a given species in each sample

```{r preprocess2d, warning=FALSE, message=FALSE}
head(get_sample(pseq, taxa_names(pseq)[1]))
```

Filter samples

```{r preprocess3, warning=FALSE, message=FALSE}
f1 <- filterfun_sample(topp(0.1))
taxa <- genefilter_sample(pseq, f1, A = round(0.5 * nsamples(pseq)))
```

Select samples by specific metadata fields:

```{r data-subsetting}
pseq.subset <- subset_samples(pseq, nationality == "AFR")
```


### Data transformations

The microbiome package provides a wrapper for standard sample/OTU transformations. For arbitrary transformations, use the transform_sample_counts function in the phyloseq package.

Log10 transformation (log(1+x) if the data contains zeroes)

```{r preprocessing-logtrans}
pseq.log <- transform_phyloseq(pseq, "log10")
```

Z transformation:

```{r preprocessing-ztrans}
pseq.zotu <- transform_phyloseq(pseq, "Z", "OTU")
```

Relative abundances (the input data needs to be in absolute scale, not logarithmic!):

```{r data10}
pseq1 <- transform_phyloseq(pseq, "relative.abundance", "OTU")
pseq2 <- transform_sample_counts(pseq, function(x) x/sum(x))
```


### Variable operations

Sample variable names

```{r preprocess4, warning=FALSE, message=FALSE}
sample_variables(pseq)
```

Pick variable values for a given variable

```{r preprocess4b, warning=FALSE, message=FALSE}
head(get_variable(pseq, sample_variables(pseq)[1]))
# .. or assigning fields to metadata:
# sample_data(GP)$human <- ..
```

### Taxa operations


Number of taxa

```{r preprocess3e, warning=FALSE, message=FALSE}
n <- ntaxa(pseq)
```


Names

```{r preprocess3f, warning=FALSE, message=FALSE}
ranks <- rank_names(pseq)
taxa <- taxa_names(pseq)
```


Prune taxa:

```{r preprocess3b, warning=FALSE, message=FALSE}
taxa <- map_levels(NULL, "Phylum", "Genus", pseq)$Bacteroidetes

# With given taxon names
ex2 <- prune_taxa(taxa, pseq)

# Taxa with positive sum across samples
ex3 <- prune_taxa(taxa_sums(pseq) > 0, pseq)
```


Subset taxa:

```{r preprocess3c, warning=FALSE, message=FALSE}
pseq <- subset_taxa(pseq, Phylum == "Bacteroidetes")
```


Filter by user-specified function values (here variance):

```{r preprocess3d, warning=FALSE, message=FALSE}
f <- filter_taxa(pseq, function(x) var(x) > 1e-05, TRUE)
```


List unique phylum-level groups: 

```{r preprocess3g, warning=FALSE, message=FALSE}
head(get_taxa_unique(pseq, "Phylum"))
```

Pick detected taxa by sample name:

```{r preprocess3gg, warning=FALSE, message=FALSE}
samplename <- sample_names(pseq)[[1]]
tax.abundances <- get_taxa(pseq, samplename)
```


Taxa sums

```{r preprocess3h, warning=FALSE, message=FALSE}
head(taxa_sums(pseq))
```


### Merging operations

Aggregate taxa to higher taxonomic levels. This is particularly useful if the phylogenetic tree is missing. When it is available, see [merge_samples, merge_taxa and tax_glom](http://joey711.github.io/phyloseq/merge.html))

```{r data24}
pseq2 <- summarize_taxa(pseq, "Phylum") 
```


Merging phyloseq objects

```{r preprocess-merge, warning=FALSE, message=FALSE, eval=FALSE}
merge_phyloseq(pseqA, pseqB)
```


### Rarification

```{r preprocess-rarify, warning=FALSE, message=FALSE}
pseq.rarified <- rarefy_even_depth(pseq)
```


### Taxonomy 

Convert between taxonomic levels (here from Genus (Akkermansia) to
Phylum (Verrucomicrobia)):

```{r phylogeny-example2, warning=FALSE, message=FALSE}
data(atlas1006)
pseq <- atlas1006
m <- map_levels("Akkermansia", "Genus", "Phylum", tax_table(pseq))
print(m)
```


### Metadata

Visualize frequencies of given factor (sex) levels within the
indicated groups (group):

```{r phylogeny-example3, warning=FALSE, message=FALSE, fig.width=5, fig.height=3}
data(dietswap)
res <- plot_frequencies(sample_data(dietswap), "group", "sex")
print(res$plot)
kable(res$data, digits = 2)
```

