### Probe summarization

Summarize (preprocessed) oligo-level data into phylotype level; examples with simulated data; see [read.profiling](reading) to use your own data. We use and recommend the [Robust Probabilistic Averaging (RPA)](https://github.com/antagomir/RPA/wiki) for probe summarization. <!--Preprocessing of HITChip data is based on RPA [`r citep(citation("RPA"))`; `r citep("10.1109/TCBB.2009.38")`]-->

```{r probelevel-ex}
library(microbiome)
data.directory <- system.file("extdata", package = "microbiome")

# Read oligo-level data (here: simulated example data)
probedata <- read.profiling(data.directory, method = "frpa")$probedata

# Read phylogeny map
# NOTE: use phylogeny.filtered for species/L1/L2 summarization
# Load taxonomy from output directory
taxonomy <- GetPhylogeny("HITChip", "filtered")

# Summarize oligos into higher level phylotypes
dat <- summarize_probedata(
                 probedata = probedata,
		 taxonomy = taxonomy, 
                 method = "rpa",
		 level = "species")
```


### Retrieve probe-level data

Get probes for each probeset:

```{r probelevel-ex2}
sets <- retrieve.probesets(phylogeny.info, level = "species", name = NULL)
```


Get probeset data matrix/matrices:

```{r probelevel-ex3}
set <- get.probeset("Actinomyces naeslundii", "species",
       		    taxonomy, probedata, log10 = TRUE)
```





