---
title: "microbiome vignette"
author: "Leo Lahti and Jarkko Salojarvi"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    number_sections: true
    theme: united
    highlight: pygments
---
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial}
  %\usepackage[utf8]{inputenc}
-->


microbiome R package
===========

The microbiome package contains general-purpose tools for
microarray-based analysis of microbiome profiling data sets. 

## Installation

### Installing and loading the experimental development version

```{r install2, eval=FALSE}
install.packages("devtools")
library(devtools)
install_github("microbiome/microbiome")
```

### Loading the package

```{r loading, eval=TRUE}
library(microbiome)  
```


## Examples

### PeerJ example data set

An example data set from Lahti et al. [PeerJ 1:e32, 2013](https://peerj.com/articles/32/) concerns associations between human intestinal microbiota and blood serum lipids. Load the data in R:

```{r exdata}
library(microbiome)
data(peerj32)
names(peerj32)
```

### Load example data

Load simulated example data of the human gut microbiota. Note that
with HITChip, fRPA is the recommended preprocessing method (kindly
cite [this
article](http://www.computer.org/csdl/trans/tb/2011/01/ttb2011010217-abs.html)). 


```{r data, eval=TRUE, message=FALSE, warning=FALSE}
# Define data path (you can replace data.directory with your own path)
data.directory <- system.file("extdata", package = "microbiome")

# Read HITChip data matrix (genus-level (L2) log10 values)
level <- "L2"
method <- "frpa"
genus.data <- read.profiling(level = level, 
	     		       method = method, 
              		       data.dir = data.directory, 
	      	       	       log10 = TRUE)  

# Read HITChip probe level data (absolute values - no log10)
oligo.data <- read.profiling(level = "oligo", 
                             data.dir = data.directory, 
			     log10 = FALSE)  

# Probe-taxon mapping table
phylogeny.info <- read.profiling(level = "phylogeny.full", 
                           	 data.dir = data.directory)

# Phylogeny that is used to summarize the probes to phylotype/genus/phylum levels
phylogeny.info.filtered <- read.profiling(level = "phylogeny.filtered", 
                           	 data.dir = data.directory)
```



### Read metadata

An easy way to provide sample metadata is to create a tab-separated metadata file. You can create the file in Excel and export it to tab-separated csv format. The standard (and self-explanatory) field names include 'sampleID', 'time', 'subjectID', 'group', 'gender', 'diet', 'age'. You can leave these out or include further fields. See this [example file](https://raw.github.com/microbiome/microbiome/master/inst/extdata/metadata.xls). Read the metadata with:

```{r meta, eval=TRUE, message=FALSE, warning=FALSE}
# Read simulated example metadata
library(gdata)
metadata.file <- paste(data.directory, "/metadata.xls", sep = "")
metadata <- read.xls(metadata.file, as.is = TRUE)
rownames(metadata) <- metadata$sampleID
```


## Usage Examples

### Diversity estimation

```{r diversity-example, warning=FALSE, message=FALSE}
# Determine detection threshold as the top 80 percent quantile
# of the data
det.th <- quantile(oligo.data, 0.8)

# Visualize the detection threshold (at log10 scale for clarity)
plot(density(log10(oligo.data))); abline(v = log10(det.th), main = "Detection threshold", xlab = "Abundance (Log10)", ylab = "Frequency")

# Calculate richness. 
# This indicates how many oligos are present in each sample
# (exceed the detection threshold)
ri <- colSums(oligo.data > det.th)

# Diversity using the vegan package
# NOTE: data needs to be in absolute scale, not logarithmic
di <- vegan::diversity(t(oligo.data), index = "shannon")

# Pielou's evenness is S/ln(R) w.r.t. given detection threshold
# NOTE: here we use detection threshold for diversity as well because
# the exact same data has to be used for diversity and richness calculation,
# and for richness calculation the detection threshold needs to be set anyway
# Diversity can be as such calculated also without threshold (see above)
# but that gives somewhat different result.
oligo.data2 <- oligo.data - det.th # NOTE: absolute (not log) scale data
S <- vegan::diversity(t(oligo.data2), index = "shannon")
R <- colSums(oligo.data2 > 0)
ev <- S/log(R)

# Combine all into a single table
divtab <- cbind(richness = ri, evenness = ev, diversity = di)
head(divtab)
```


Calculate diversities within each phylum-level group

```{r L1div, warning=FALSE, message=FALSE}

L1.groups <- unique(phylogeny.info.filtered$L1) # List all L1 groups
L1.diversities <- NULL # initialize
for (phylum in L1.groups) {
  # List all probes for the given phylum
  probes <- levelmap(phylum, "L1", "oligoID", phylogeny.info = phylogeny.info.filtered)[[phylum]]
  # Check diversity within this phylum
  di <- vegan::diversity(t(oligo.data[probes,]), index = "shannon")
  # Add to the table
  L1.diversities <- cbind(L1.diversities, di)
}
# Name the columns
colnames(L1.diversities) <- L1.groups
```


### Phylogeny

Map phylotypes between hierarchy levels:

```{r phylogeny-example2, warning=FALSE, message=FALSE}
phylogeny.info <- GetPhylogeny("HITChip")
m <- levelmap(phylotypes = NULL, 
              level.from = "species", 
	      level.to = "L2", 
	      phylogeny.info = phylogeny.info)
```


### Diversity boxplot

Diversity boxplot for selected samples:

```{r diversity-example2, warning=FALSE, message=FALSE}
group <- metadata$group
names(group) <- metadata$sampleID
my.samples <- names(group)
boxplot(di[my.samples]  ~ group[my.samples], las = 1)
```

### Estimating relative abundancies

Estimate relative abundance of the taxa in each sample. Note: the
input data set needs to be in absolute scale (not logarithmic).

```{r diversity-example6}
rel <- relative.abundance(oligo.data, det.th = NULL)
```


### Core microbiota

Determine common core microbiota, following the [blanket
analysis](http://onlinelibrary.wiley.com/doi/10.1111/j.1469-0691.2012.03855.x/abstract):
 
```{r core-example, message=FALSE, warning=FALSE}
core <- createCore(t(peerj32$microbes))
```

Visualizing core microbiota:

```{r core-example2, fig.width=8, fig.heigth=8}
# Core 2D visualization
tmp <- Core2D(core)

# Core heatmap
tmp <- core_heatmap(t(peerj32$microbes))
```

### Cross-correlation example

```{r heatmap}
dat1 <- peerj32$lipids # Lipids (44 samples x 389 lipids)
dat2 <- peerj32$microbes # Microbiota (44 samples x 130 bacteria)
meta <- peerj32$meta

correlations <- cross.correlate(dat1, dat2, 
                        method = "bicor", 
			mode = "matrix", 
                        n.signif = 1, 
			p.adj.threshold = 0.05, 
                        p.adj.method = "BH")


correlation.table <- cmat2table(correlations)
head(correlation.table)
```

### Prevalence of taxonomic groups

```{r prevalence}
# List prevalence measure for each group using detection threshold of 2
# Sort the taxa by prevalence
head(prevalence(peerj32$microbes, 2, sort = TRUE))

# Just list the names of taxa that are present over abundance threshold 2
# in over 20 percent of the samples:
prevalent.taxa <- list_prevalent_groups(peerj32$microbes, 2, 0.2)
```


### ROC analysis

A basic example of ROC/AUC analysis with simulated random data.

```{r roc-example, warning=FALSE}
# Define two sample groups for demonstration purpose
g1 <- sample(colnames(oligo.data), 10)
g2 <- setdiff(colnames(oligo.data), g1)

# Compare the two groups with t-test
pvalues <- c()
for (tax in rownames(oligo.data)) {
  pvalues[[tax]] <- t.test(oligo.data[tax, g1], oligo.data[tax, g2])$p.value
}

# Order the taxa based on the p-values
ordered.results <- names(sort(pvalues))

# Assume there are some known true positives (here just randomly picked for demonstration)
true.positives <- sample(rownames(oligo.data), 10)

# Overall ROC analysis (this will give the cumulative TPR and FPR along the ordered list)
res <- roc(ordered.results, true.positives)

# Calculate ROC/AUC value
auc <- roc.auc(ordered.results, true.positives)
print(auc)

# Plot ROC curve
roc.plot(ordered.results, true.positives, line = TRUE)
```



### Licensing and Citations

This work can be freely used, modified and distributed under the 
[Two-clause FreeBSD license](http://en.wikipedia.org/wiki/BSD\_licenses).

Kindly cite the work as 'Leo Lahti and Jarkko Salojarvi
(2014). microbiome R package. URL: http://microbiome.github.com'.


### References

The package utilizes tools from a number of other CRAN and
Bioconductor extensions, including ade4, df2json, rjson, fastcluster,
ggplot2, MASS, methods, minet, mixOmics, plyr, qvalue, RCurl,
reshape2, RPA, vegan, and WGCNA. We thank all authors for these
contributions:

 * N. Caballero (2013). [df2json: Convert a dataframe to JSON](http://CRAN.R-project.org/package=df2json) 

 * A. Couture-Beil (2013). [rjson: JSON for R](http://CRAN.R-project.org/package=rjson) 

 * A. Dabney, John D. Storey and with assistance from Gregory R. Warnes. qvalue: Q-value estimation for false discovery rate control. 

 * S. Dray and A. B. Dufour, (2007): The ade4 package: implementing the duality diagram for ecologists. Journal of Statistical Software. 22(4): 1-20.

 * S. Dejean et al. (2013). [mixOmics: Omics Data Integration Project](http://CRAN.R-project.org/package=mixOmics) 

 * L. Lahti et al. A fully scalable online-preprocessing algorithm for short oligonucleotide microarray atlases. [NAR 41(10):e110, 2013](http://nar.oxfordjournals.org/content/41/10/e110) 

 * L. Lahti et al. Analysis of Probe Reliability in Differential Gene Expression Studies with Short Oligonucleotide Arrays. [TCBB/IEEE 8(1):217-225, 2011](http://www.computer.org/portal/web/csdl/doi/10.1109/TCBB.2009.38)

 * L. Lahti et al. Associations between the human intestinal microbiota, Lactobacillus rhamnosus GG and serum lipids indicated by integrated analysis of high-throughput profiling data. [PeerJ 1:e32, 2013](http://dx.doi.org/10.7717/peerj.32).

 * D. T. Lang (2013). [RCurl: General network (HTTP/FTP/...) client interface for R](http://CRAN.R-project.org/package=RCurl) 

 * P. Langfelder and S. Horvath, WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics 2008, 9:559 

 * P. Langfelder, S. Horvath (2012). Fast R Functions for Robust Correlations and Hierarchical Clustering. [Journal of Statistical Software, 46(11), 1-17](http://www.jstatsoft.org/v46/i11/)

 * P. E. Meyer, Frederic Lafitte and Gianluca Bontempi (2008). MINET: An open source R/Bioconductor Package for Mutual Information based Network Inference. [BMC Bioinformatics](http://www.biomedcentral.com/1471-2105/9/461)

 * D. Mullner (2013). fastcluster: Fast Hierarchical, Agglomerative Clustering Routines for R and Python. [Journal of Statistical Software, 53(9), 1-18](http://www.jstatsoft.org/v53/i09/)

 * Jari Oksanen et al. (2013). [vegan: Community Ecology Package](http://CRAN.R-project.org/package=vegan) 

 * R Core Team (2013). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. [ISBN 3-900051-07-0](http://www.R-project.org/)

 * W. N. Venables and B. D. Ripley (2002) Modern Applied Statistics with S. Fourth Edition. Springer, New York. ISBN 0-387-95457-0

 * H. Wickham (2007). Reshaping Data with the reshape Package. [Journal of Statistical Software, 21(12), 1-20](http://www.jstatsoft.org/v21/i12/)

 * H. Wickham. ggplot2: elegant graphics for data analysis. Springer New York, 2009. 

 * H. Wickham (2011). The Split-Apply-Combine Strategy for Data Analysis. [Journal of Statistical Software, 40(1), 1-29](http://www.jstatsoft.org/v40/i01/)


### Session info

This vignette was created with

```{r sessioninfo}
sessionInfo()
```




