---
title: "Introduction to the microbiome R package"
author: "Leo Lahti, Sudarshan Shetty, et al."
bibliography: 
- bibliography.bib
date: "`r Sys.Date()`"
output:
  rmarkdown::pdf_document:
    toc: true
  rmarkdown::md_document:
    toc: true    
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{microbiome R package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
require(knitcitations)
# cleanbib()
# options("citation_format" = "pandoc")
bib <- read.bibtex("bibliography.bib")

# Can also write math expressions, e.g. 
# $Y = X\beta + \epsilon$, footnotes^[A footnote here.]
# > "He who gives up [code] safety for [code] speed deserves neither."
# ([via](https://twitter.com/hadleywickham/status/504368538874703872))
```

## Introduction

The [microbiome R package](http://microbiome.github.io/microbiome) facilitates exploration and analysis of microbiome profiling data, in particular 16S taxonomic profiling. Here we provide a very concise overview to the package functionality using example data sets from published studies [@lahti14natcomm, @Lahti13provasI, @OKeefe15]. We have intentionally omitted many details in order to provide an overview. A full tutorial is available [on-line](http://microbiome.github.io/microbiome). 

The microbiome R package has been utilized in recent publications [@salonen14, @Faust15, @Shetty2017], and is released under the [Two-clause FreeBSD license](http://en.wikipedia.org/wiki/BSD\_licenses). The package extends the independent [phyloseq](http://joey711.github.io/phyloseq/import-data) class structure and [tools](http://joey711.github.io/phyloseq/).

Input from the user community is very welcome via the [issue tracker](https://github.com/microbiome/microbiome/issues) or [pull requests](http://microbiome.github.io/microbiome/Contributing.html). See the [Github site](https://github.com/microbiome/microbiome) for source code and other details. 

Kindly cite the work as follows: "Leo Lahti [et al.](https://github.com/microbiome/microbiome/graphs/contributors) (2017). Tools for microbiome analysis in R. Microbiome package version `r sessionInfo()$otherPkgs$microbiome$Version`. URL: [http://microbiome.github.com/microbiome](http://microbiome.github.com/microbiome). 


## Installation

To install microbiome package (the latest development version) in R, use

```{r microbiomeinstall, message=FALSE, warning=FALSE, eval=FALSE}
library(devtools) # Load the devtools package
install_github("microbiome/microbiome") # Install the package
```

Then load the package in R

```{r loading, eval=TRUE, message=FALSE}
library(microbiome)  
```


## Data

A [phyloseq](http://joey711.github.io/phyloseq) data object typically contains an OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and a phylogenetic tree (relations between the taxa). 


### Example data

HITChip Atlas data set [Lahti et al. Nat. Comm. 5:4344, 2014](http://www.nature.com/ncomms/2014/140708/ncomms5344/full/ncomms5344.html)) contains 130 genus-level taxonomic groups across 1006 western adults. Get example data with

```{r atlasdata}
data(atlas1006) 
```

A two-week diet swap study between western (USA) and traditional (rural Africa) diets, reported in [O'Keefe et al. Nat. Comm. 6:6342, 2015](http://dx.doi.org/10.1038/ncomms7342)

```{r dietswapdata}
data(dietswap) 
```

A gut microbiome profiling data set, which includes parallel profiling of intestinal microbiota versus blood metabolites from [Lahti et al. PeerJ 1:e32, 2013](https://peerj.com/articles/32/) to characterize associations between human intestinal microbiota and blood serum lipids. 

```{r peerjdata}
data(peerj32) # Data from https://peerj.com/articles/32/
```


### Data import

The [tutorial](http://microbiome.github.io/microbiome/Data.html) describes how to import data from standard formats (Mother, BIOM, CSV).


### Data manipulation

A phyloseq object can be subsetted, filtered, aggregated, transformed, and otherwise manipulated with the [phyloseq](http://joey711.github.io/phyloseq/) and [microbiome](http://microbiome.github.io/microbiome/Preprocessing.html) tools. The microbiome package provides a wrapper for many standard transformations such  as Z, centered log-ratio, hellinger, log10, and other transformations. 

To convert absolute counts to compositional (relative) abundances, for instance, use

```{r compositional}
pseq.comp <- transform(dietswap, "compositional")
```



## Diversity and other ecosystem indices 

Standard ecosystem state variables include richness, evenness, alpha diversity, dominance, and rarity. The function `global` calls these indicators with default parameters. For further options and [beta diversity](http://microbiome.github.io/microbiome/Betadiversity.html) quantification, see [tutorial](http://microbiome.github.io/microbiome/Diversity.html). 

```{r global, warning=FALSE, message=FALSE, results="asis"}
g <- global(atlas1006, index = "gini")
```


Regression curve with smoothed error bars based on the [Visually-Weighted Regression](http://www.fight-entropy.com/2012/07/visually-weighted-regression.html) can be used to visualize sample variables.

```{r variability-regression, message=FALSE, eval=TRUE, fig.width=10, fig.height=5, warning=FALSE, dev="CairoPNG"}
plot_regression(diversity ~ age, meta(atlas1006))
```


## Core microbiota analysis

Population frequencies, **prevalence**, of the taxonomic groups exceeding a given detection thresholld can be calculated with

```{r core-prevalence2}
p <- prevalence(dietswap, detection = 0, sort = TRUE)
```

The core microbiota refers to the set of taxa that are detected in a remarkable fraction of the population above a given abundance threshold [@Jalanka-Tuovinen11, @Salonen12cmi]. The core subset can be sliced from a phyloseq object as follows.

```{r core-data, message=FALSE, warning=FALSE}
dietswap.core <- core(dietswap, detection = 0, prevalence = 50/100)
```

Similar functions are available also for rare and variable taxa, see the microbiome [tutorial](http://microbiome.github.io/microbiome/Core.html) for a full description.

To visualize the core as in [@Shetty17], use

```{r corevisu, fig.width=8, fig.heigth=18, fig.show='hold', out.width = '300px'}
library(ggplot2)
p <- plot_core(transform(dietswap.core, "compositional"), 
    plot.type = "heatmap", 
    colours = gray(seq(0,1,length=5)),
    prevalences = seq(.05, 1, .05), 
    detections = 10^seq(log10(1e-3), log10(.2), length = 10)) +
    xlab("Detection Threshold (Relative Abundance (%))")
print(p)    
```


## Microbiome composition


Composition heatmap: Z-transformed abundances for the core taxa

```{r compheat, message=FALSE, warning=FALSE, fig.height=10, fig.width=10}
# Focus on the core taxa
pseq.core <- core(transform(dietswap, "compositional"), 
            detection = .2/100, prevalence = 50/100)
# Show Z-transformed abundances
tmp <- plot_composition(pseq.core, plot.type = "heatmap", transform = "Z", 
            mar = c(6, 13, 1, 1), sample.sort = "nationality")
```   


Composition barplot
        
```{r compbar, message=FALSE, warning=FALSE, fig.height=10, fig.width=10}
# Compare relative abundances of the core taxa
pseq.core <- transform(pseq.core, "compositional")

# Visualize
p <- plot_composition(pseq.core, plot.type = "barplot", sample.sort = "neatmap")
print(p)
```

Visualize the microbiome landscape [@Shetty2017]. A number of other [ordination methods](http://microbiome.github.io/microbiome/Ordination.html) (PCoA, NMDS etc) are available as well.

```{r landscape4, message=FALSE, warning=FALSE, fig.width=8, fig.height=6, out.width="400px"}
# Project the samples with the given method and dissimilarity measure. 
# Ordinate the data; note that some ordinations are sensitive to random seed
# "quiet" is used to suppress intermediate outputs
set.seed(423542)
quiet(proj <- get_ordination(pseq.core, "NMDS", "bray"))

# random seed affects the exact ordination
p <- plot_landscape(proj[, 1:2], col = proj$nationality, legend = T)
print(p)
```


Bagged RDA is a robust version compared to the standard RDA. Fit bagged (bootstrap aggregated) RDA on a phyloseq object.

```{r rda, warning=FALSE, message=FALSE, eval=TRUE}
pseq <- peerj32$phyloseq # phyloseq data

# Core taxa to speed up examples
pseq <- core(pseq, detection = 10^2, prevalence = 95/100)
pseq.trans <- microbiome::transform(pseq, "hell") # Hellinger transform

# In any real study, use bs.iter = 100 or higher
# to achieve meaningful benefits from the bagged version.
# In this example we use bs.iter = 2 just to speed up the
# example code for educational purposes
res <- rda_bagged(pseq.trans, "group", bs.iter=2)
```

Visualizing bagged RDA:

```{r rda6, warning=FALSE, message=FALSE, fig.width=8, fig.height=8, eval=TRUE}
plot_rda_bagged(res)
```



### Association heatmaps

To exemplify how to cross-correlate two data sets, we use data from microbiome and blood serum lipids study ([PeerJ 1:e32](https://peerj.com/articles/32/)).

```{r heatmap-crosscorrelate4, message=FALSE, warning=FALSE, fig.keep='none'}
data(peerj32)
otu <- peerj32$microbes 
lipids <- peerj32$lipids 

# Define data sets to cross-correlate
x <- log10(otu) # OTU Log10 (44 samples x 130 genera)
y <- as.matrix(lipids) # Lipids (44 samples x 389 lipids)

# Cross correlate data sets and return a table
correlation.table <- associate(x, y, method = "bicor", mode = "table", p.adj.threshold = 0.05, n.signif = 1)

kable(head(correlation.table))
```

Let us rearrange the data and plot the heatmap and highlight significant correlations with stars as in [@Lahti13provasI].

```{r heatmap-example-stars3, fig.width=12, fig.height=8, message=FALSE, warning=FALSE}
heat(correlation.table, "X1", "X2", fill = "Correlation", star = "p.adj", p.adj.threshold = 0.05) 
```


## Bistability and tipping elements

Certain microbial groups exhibit bi-stable abundance distributions [see e.g. @Lahti2014]. The microbiome package provides tools to [quantify stability and bimodality](http://microbiome.github.io/microbiome/Stability.html) in abundance data. Tools are provided also for related visualizations.

Let use Dialister as an example as this has been suggested to exhibit bistable dynamics [@Lahti2014] in healthy western adults. 

```{r stability-variationplot2, message=FALSE, warning=FALSE, fig.show='hold', out.width="430px"}
# Use relative abundances and plot subject variation 
# during the follow-up period
pseq <- transform(atlas1006, "compositional")
pv <- plot_tipping(pseq, "Dialister", tipping.point = 0.004)
print(pv)

# Show the bimodal population distribution
pb <- hotplot(pseq, "Dialister", tipping.point = 0.004)
print(pb)
```


## Other tools

The [on-line tutorial](http://microbiome.github.io/microbiome) provides additional tools and examples, for instance for [microbiome heatmaps](Composition.html), group-wise community comparisons](Comparisons.html), and [microbial network analysis](Comparisons.html).


### Acknowledgements

Thanks to all [contributors](https://github.com/microbiome/microbiome/graphs/contributors). Financial support has been provided by Academy of Finland (grants 256950 and 295741), [University of Turku](http://www.utu.fi/en/Pages/home.aspx), Department of Mathematics and Statistics, [VIB lab for Bioinformatics and (eco-)systems biology](http://www.vib.be/en/research/scientists/Pages/Jeroen-Raes-Lab.aspx), VIB/KULeuven, Belgium, [Molecular Ecology group](http://www.mib.wur.nl/UK/), Laboratory of Microbiology, Wageningen University, Netherlands, and [Department of Veterinary Bioscience](http://www.vetmed.helsinki.fi/apalva/index.htm), University of Helsinki, Finland. This work relies on the independent [phyloseq](https://github.com/joey711/phyloseq) package and data structures for R-based microbiome analysis developed by Paul McMurdie and Susan Holmes. This work also utilizes a number of independent R extensions, including ade4 [@ade4], dplyr [@dplyr], ggplot2 [@ggplot2], MASS [@MASS], moments [@moments], phyloseq [@phyloseq], RColorBrewer [@RColorBrewer], scales [@scales], stats [@stats], tidyr [@tidyr], and vegan [@vegan].



# References


```{r, echo=FALSE, message=FALSE}
#embed citations, for example: `r citep(bib[["lahti14natcomm"]])`
#embed citations, for example2: @lahti14natcomm
#Cite with DOI: `r citep("10.1890/11-0011.1")`
#Cite URL `r citep("http://knowledgeblog.org/greycite")`
#For automated markdown citations, check [this](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html), [this](https://github.com/cboettig/knitcitations), and [this](http://www.carlboettiger.info/2012/03/24/citations-in-markdown-using-knitr.html). 
#write.bibtex(file="references.bib")
```
```{r, echo=FALSE, message=FALSE, results="asis"}
bibliography()
```


