---
title: "microbiome vignette"
author: "Leo Lahti and Jarkko Salojarvi"
date: "`r Sys.Date()`"
bibliography: 
- bibliography.bib
- references.bib
output: 
  md_document:
    variant: markdown_github
---
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial}
  %\usepackage[utf8]{inputenc}
-->

```{r, echo=FALSE, message=FALSE, warnings=FALSE}
# Handle citations
require(knitcitations)
cleanbib()
#options("citation_format" = "pandoc")
bib <- read.bibtex("bibliography.bib")
#opts_chunk$set(fig.width=4, fig.height=3, par=TRUE, out.width='2in', fig.pos='H')
#knitr::opts_chunk$set(fig.path = "figure/")
```

microbiome R package
===========

Tools for microbiome analysis in R (`r citep(bib[["rcore13"]])`).


### Getting started

* [Installation](Template.md) 
* [Example data](Data.md)
* [Data manipulation](Preprocessing.md)


### Visualization and related tools

* [Community composition](Composition.md)
* [Heatmaps](Heatmap.md)
* [Networks](Networks.md)
* [Ordination (PCA, PCoA, NMDS etc.)](Ordination.md)
* [Density](Density.md)
* [Interactive](Interactive.md)


### Microbiota composition

* [Bistability analysis](Stability.md)
* [Variability](Variability.md) (Intra- and inter-individual 'stability')
* [Core microbiota](Core.md)
* [Diversity](Diversity.md)


### Analysis

* [Pairwise comparisons](Comparisons.md)
* [Stability, bimodality, tipping elements](Stability.md)
* [Clustering](Clustering.md)
* [Linear models](limma.md)
* [Phylogenetic microarrays](HITChip.Rmd)



### Licensing and Citations

This work can be freely used, modified and distributed under the
[Two-clause FreeBSD
license](http://en.wikipedia.org/wiki/BSD\_licenses). Kindly cite as
'Leo Lahti and Jarkko Salojarvi (2014-2016). microbiome R
package. URL: http://microbiome.github.com'.


### Dependencies

The package utilizes tools from a number of other CRAN and
Bioconductor extensions, including:

 * ade4 `r citep(citation("ade4"))`
 * df2json `r citep(citation("df2json"))`
 * dplyr `r citep(citation("dplyr"))`  
 * fastcluster `r citep(citation("fastcluster"))`
 * ggplot2 `r citep(citation("ggplot2"))`
 * minet `r citep(citation("minet"))` 
 * mixOmics `r citep(citation("mixOmics"))`
 * phyloseq `r citep(citation("phyloseq"))`
 * RPA `r citep(citation("RPA"))` 
 * reshape `r citep(citation("reshape"))` 
 * rjson `r citep(citation("rjson"))`
 * RCurl `r citep(citation("RCurl"))`
 * vegan `r citep(citation("vegan"))`
 * WGCNA `r citep(citation("WGCNA"))`


### References

```{r, echo=FALSE, message=FALSE}
#You can embed citations, for example: `r citep(bib[["lahti14natcomm"]])`
#You can embed citations, for example2: @lahti14natcomm
#Cite with DOI: `r citep("10.1890/11-0011.1")`
#Cite URL `r citep("http://knowledgeblog.org/greycite")`
#For automated markdown citations, check [this](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html), [this](https://github.com/cboettig/knitcitations), and [this](http://www.carlboettiger.info/2012/03/24/citations-in-markdown-using-knitr.html). 
#write.bibtex(file="references.bib")
```

```{r, echo=FALSE, message=FALSE, results='asis'}
bibliography()
```

### Session info

This vignette was created with

```{r sessioninfo}
sessionInfo()
```




