%\VignetteIndexEntry{microbiome}
%The above line is needed to remove a warning in R CMD check
\documentclass[a4paper]{article}
\usepackage[utf8]{inputenc}
\usepackage{amsmath,amsthm,amsfonts}
\usepackage{graphicx}
%\usepackage[T1]{fontenc}
%\usepackage[latin1]{inputenc}
%\usepackage[authoryear,round]{natbib}
\usepackage[numbers]{natbib}
\usepackage{hyperref}
\usepackage{Sweave}
\usepackage{float}

\setlength{\textwidth}{450pt}
\setlength{\oddsidemargin}{20pt}
\setlength{\evensidemargin}{20pt}
%\setlength{\marginparwidth}{55pt}

\setlength{\parindent}{0mm}
\setlength{\parskip}{2mm}
\setlength{\topmargin}{5pt}
\addtolength{\textheight}{50pt}
%\textwidth=6.2in
%\textheight=8.5in
%\oddsidemargin=.1in
%\evensidemargin=.1in
%\headheight=-.3in

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\title{microbiome\\R package and protocols}
\author{Contact: Leo Lahti\footnote{Wageningen University, Netherlands; leo.lahti@iki.fi}, Jarkko
Saloj{\"a}rvi\footnote{University of Helsinki, Finland; jarkko.salojarvi@helsinki.fi}}

%\hyphenation{}

\begin{document}

\maketitle

\section{Introduction}

This is a development package (alpha release) which includes standard
R routines for preprocessing, analysis and visualization of microbiota
profiling data. The package is jointly developed by researchers from
the Molecular Ecology group, Laboratory of Microbiology, Wageningen
University, Netherlands, and Gut Group, Department of Veterinary
Bioscience, University of Helsinki, Finland. 

{\bf Contact:} microbiome-admin@googlegroups.com


\section{Profiling script}

\subsection{Short instructions}

The profiling script is executed with a single R command which depends
on the Chip. The program will then guide you through the data and
parameter selection. The following examples require that the phylochip
MySQL database is installed on the system. For HITChip, use:

<<profiling,results=hide,eval=FALSE>>=
library(microbiome) 
params <- run.profiling.script("root", "array", "Phyloarray") 
@

For PITChip and MITChip replace dbname with Phyloarray\_PIT or Phyloarray\_MIT, respectively.

%params <- run.profiling.script(dbuser, dbpwd, dbname = "pitchipdb")  # PITChip Linux

\subsection{Detailed protocol for the Windows database computer}

Profiling is conducted with R program using ready-made scripts. To
analyse chip data, carry out the following steps on the WUR database
computer. This will take you through parameter selection steps; the
default options are highlighted, you can keep the defaults unless
there are specific reasons for changes.

\begin{enumerate}

\item Open R (currently version Rx64 2.12.2 at WUR)

\item Type in R the following commands to analyze HITChip data (for
PITChip and MITChip, replace dbname (Phyloarray) with Phyloarray\_PIT
(old generation), pitchipdb (new PITChip), or Phyloarray\_MIT,
respectively):

<<profiling,results=hide,eval=FALSE>>=
library(microbiome)
params <- run.profiling.script("root", "array", "Phyloarray")
@


\item Choose whether to input parameters manually, or read them from an R file.

\item Select the folder for output files produced by the Script.

\item Select the data for the analyses: (a) New data from database
(when starting the analysis): (b) Existing data in the workspace (to
reanalyse data that you were working during this R session).

\item Select Study
\item Select Samples (two or more)
\item Select phylogeny used for profiling if multiple options are available for your Chip (standard: 16S)
\item Select whether to remove ("Yes") or keep ("No") non-specific oligos. Keeping non-specific oligos by default.
\item Select scaling method (standard selection is Minimum/Maximum. For more information, see below)
\item Select default plotting parameters or define them by yourself: 
  \begin{itemize}
    \item Colour scale (standard: white/blue)
    \item Taxonomic level to display in the graph (standard: level 1)
    \item Clustering metric (standard: Pearson correlation coefficient)
    \item Whether to display tree in graph
    \item Percentage of tree height to total figure (standard 12)
    \item Font size (standard 12)
 \end{itemize}
\item Wait until the calculations are completed (typically within 5-30 minutes). Please do not interrupt the calculations.
\item Your profiling is finished. You get 2 versions of clusters: raw and log 10. To save these, right click on the picture and save as metafile. 
\item Output files are in your selected folder (from step 4), including:

  \begin{itemize}

    \item Profiling log (txt file)

    \item Phylotypes x samples data matrices at absolute scale for two alternative oligo summarization methods
    (sum, rpa) and taxonomic levels (species, level 2, level 1). For MITChip, also level 0 is provided. The data matrices are
    in the *.tab files which can also be opened with Excel. By default, you
    can use the sum version. For details, see Probe summarization
    section below. If you need logarithmized scale, this has to be done separately. 

    \item oligoprofile.tab (oligos vs. samples data matrix in the original scale)

    \item phylogenyinfo.tab 

    \item oligoprofileClustering.png

  \end{itemize}

\item The “Existing data in the workspace” has been removed. The
      run.profiling.script is provided solely for preprocessing
      purposes and it should not be mixed with downstream analysis of
      the data. To preprocess the data with different parameters,
      rerun the complete run.profiling.script

\end{enumerate}

\subsubsection{Scaling: additional notes}

Scaling is used to correct for technical differences in sample
intensities. The standards options are available, corresponding to
different background assumptions:

\begin{itemize}

\item {\bf Min/max:} The default option. Aims to normalize the samples
to comparable scales without affecting the phylotype distribution
within a sample. Particularly useful when there are considerable
systematic differences between samples, for instance due to
antiobiotic treatment. Sets the minimum and maximum values of each
sample to the same value by shifting and scaling the samples
accordingly. These values are selected as the 0.5\% 99.5\% quantiles
of the data to improve robustness.

\item {\bf Quantile:} Quantile normalization may be more efficient in
removing technical biases from the data but it assumes that the
overall phylotype distribution is approximately same for the different
samples (ie. there are no antiobiotic treatments, and the
samples/subjects are also otherwise similar). The quantile
normalization forces the same phylotype distribution on all samples,
which is calculated based on the average over all samples.

\item {\bf None:} skip normalization 

\end{itemize}

\subsubsection{Probe summarization: additional notes}

Each phylotype is targeted by multiple oligos. The probe-level
information is summarized to obtain the final abundancy estimates for
the phylotypes. Three different summary measures are included in the ouput:

\begin{itemize}

\item {\bf sum:} Use this as default option. Phylotype abundancy is
calculated as a sum of the matching oligos. Cross-hybridizing oligos
that target multiple phylotypes are downweighted by the number of
targets.

\item {\bf rpa:} Robust Probabilistic Averaging method quantifies
oligo-specific noise and downweights more noisy oligos.  Oligos with
multiple targets are also downweighted automatically.

\item {\bf ave:} Average of the oligos. Does not take into account
cross-hybridization of the oligos with multiple targets. Provided for
backward compatibility.

\end{itemize}


\subsection{Reading the profiling results in R}

Assuming you have stored the output of the profiling script in
'datadirectory/' directory, you can read the phylochip data as follows
(you can also choose other levels and summarization methods). If you
do not provide data directory, the GUI will ask for the file (and also
override the level and method if these do not match with your selected
file):

<<reading,results=hide,eval=FALSE>>=
# Preprocessed phylotype-level data in log10 domain
dat <- read.profiling("level2", "rpa", "datadirectory/") 

# Oligo level data in original domain
oligodata <- read.profiling("datadirectory/", "oligo") 

# Oligo-phylogeny mappings
phylogeny <- read.profiling("datadirectory/", "phylogeny")
@


\subsection{Diversity indices}

Estimate diversity, richness, and evenness for the samples:
<<diversity,results=hide,eval=FALSE>>=
div <- estimate.diversity(read.profiling("oligo"), diversity.index = "shannon", det.th = NULL)
@

Calculate and plot diversity indices:
<<diversityplot,results=hide,eval=FALSE>>=
div <- PlotDiversity(read.profiling("oligo"), "shannon")
@


\subsection{Redrawing images}

To tune the figures, you can reload the R data objects from the
previous session and change arguments in the function call (see help(PlotPhylochipHeatmap)). You can
check from the 'params' object in R which values were used in the
previous session.

<<redraw,results=hide,eval=FALSE>>=
library(microbiome)
oligodata <- read.profiling("datadirectory/", "oligo")
phylogeny <- read.profiling("datadirectory/", "phylogeny")
png(filename = "oligoprofileClustering.png", width = max(trunc(ppcm*21), 
      trunc(ppcm*21*ncol(plotmat)/70)), height = trunc(ppcm*29.7))     
tmp <- PlotPhylochipHeatmap(oligodata, phylogeny)
dev.off()
@



<<redraw2,results=hide,eval=FALSE>>=
oligodata <- read.profiling("datadirectory/", "oligo")
hc <- calculate.hclust(oligodata, hclust.method = "complete", metric = "correlation")
png("hclust.log10.png")
plot(hc$log10, hang=-1, main = "Hierarchical clustering, oligolevel, log10")
dev.off()
png("hclust.abs.png")
plot(hc$raw, hang=-1, main = "Hierarchical clustering, oligolevel, raw")
dev.off()
@


\section{HITChip phylogenetic tree}

Calculate HITChip phylogenetic tree for a given level using (assuming you have the two files):
<<phylo,results=hide,eval=FALSE>>=
f.phylomap <- "HCphylogeny.tab.gz"
f.phylodist <- "Sebastian_HITChip.dismat.gz"
pd <- hitchip.phylodistance("level 2", f.phylomap, f.phylodist)
@


\section{Tools}

\subsection{Stability analysis}

Give input data 'dat' as phylotypes vs. samples matrix and calculate
stability as average Pearson correlation between the samples:

<<atlas,results=hide,eval=FALSE>>=
stability <- calculate.stability(dat)$stability
@



\section{HITChip atlas}

To fetch and preprocess HITChip atlas on one go, use the following
(check MySQL database (db) parameters; expect 30-60 minutes processing
time):

<<atlas,results=hide,eval=FALSE>>=
library(microbiome)
atlas <- FetchHITChipAtlas(allowed.projects = ListAtlasProjects(), 
		  dbuser = 'username', dbpwd = 'passu', dbname = 'Phyloarray', 
		  result.path = "results")
@

\section{Package installation}

Get the package tarball from package admins and run the following
installation command in R (replace x.y.z with the current version
number and give the full file path of the tarball):

<<installation,results=hide,eval=FALSE>>=
install.packages("microbiome_x.y.z.tar.gz")
@


\section{Functionality}

Documentation coming soon. The visible functions include:

\subsubsection*{Data preprocessing and I/O}

\begin{itemize}

\item FetchHITChipAtlas: Preprocessing pipeline for the HITChip atlas with default parameters

\item pick.training.samples: Split data set into training and test samples

\item run.profiling.script: Profiling main script

\item read.profiling: read run.profiling.script output into R

\item list.mysql.projects: List projects in MySQL database

\item fetch.projects: Fetch projects from the phyloarray database

\item fetch.samples: Fetch samples from the phyloarray database

\item fetch.sample.info: Fetch sample information from HITChip atlas

\end{itemize}


\subsubsection*{Analysis}

\begin{itemize}

\item distance.matrix: correlation-based distance matrix

\item pairwise.comparisons: pairwise comparison between factor levels

\item check.wilcoxon: check.wilcoxon

\item cross.correlate: Cross-correlate input variables

\item calculate.stability: calculate.stability

\item estimate.diversity: Estimate diversities for each sample (column)

\item make.abundancy.table: Calculate abundancies

\item nmf.R, lm.R: JSa

\item Phylogeneticenrichments: higher phylogenetic level enrichments

\item calculate.hclust: Calculate hierarchical clustering for standard selections in profiling script

\item project.data: Project high-dimensional data on two-dimensional plane by various methods

\end{itemize}


\subsubsection*{Utilities}

\begin{itemize}

\item include.fig; include.figs: brew help functions

\item get.phylogeny: from the MySQL db

\item levelmap: map phylotypes between hierarchy levels

\item list.probes: list probes corresponding to a given phylogenetic level

\item retrieve.probesets: Description: List probes for each probeset

\item get.probeset:  Description: get probeset data matrix

\item impute; Impute missing values from a Gaussian. 

\item Strip: Strip string i.e. remove spaces from the beginning and end

\item esort: Sort data frame dd by columns like: esort(dd, -z, b)

\item lower.triangle: Get lower triangle of a square matrix as a vector

\end{itemize}


\subsubsection*{Visualization}

\item PlotMatrix: matrix visualizations

\item PlotDiversity: Plot and save diversity indices

\item PlotPhylochipHeatmap: Plots heatmap of the oligo profiles together with phylotype grouping and sample clustering

\item theme_bottom_border: ggplot2 theme

\end{itemize}



\subsubsection*{Experimental}

\begin{itemize}

\item top.barplots: Barplot of top findings from pairwise.comparisons function

\item PlotTopComparisons: Visualize top findings from pairwise.comparisons 

\end{itemize}


\section{Citing the package}

TBA

\section{Version information}
<<details>>=
sessionInfo()
@


%\bibliographystyle[numbers]{natbib} 
%\begin{thebibliography}{1}
%\bibliographystyle{abbrv}
%\bibliography{my.bib}

\end{document}
