<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HITChip tools • microbiome</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">microbiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>HITChip tools</h1>
            
          </div>

    
    
<div class="contents">
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial - hitchip}
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}  
-->
<div id="hitchip-and-other-phylogenetic-microarrays" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#hitchip-and-other-phylogenetic-microarrays" class="anchor"> </a></body></html>HITChip and other phylogenetic microarrays</h2>
<ul>
<li><a href="https://github.com/microbiome/HITChipDB/blob/master/index./index.html">Extracting data from HITChip database</a></li>
<li><a href="Probelevel.html">Probe level studies (phylogenetic microarrays)</a></li>
</ul>
<div id="importing-hitchip-data-to-phyloseq-format" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#importing-hitchip-data-to-phyloseq-format" class="anchor"> </a></body></html>Importing HITChip data to phyloseq format</h3>
<p>Define the data folder.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define example data path (replace here data.directory with your own path)</span>
<span class="kw">library</span>(microbiome)
data.directory &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"microbiome"</span>)
<span class="kw">print</span>(data.directory)</code></pre></div>
<pre><code>## [1] "/home/lei/Rpackages/microbiome/inst/extdata"</code></pre>
</div>
</div>
<div id="installing-hitchipdb-package" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#installing-hitchipdb-package" class="anchor"> </a></body></html>Installing HITChipDB package</h2>
<p>The HITChipDB package contains additional routines to fetch and preprocess HITChip (or MIT/PITChip) data from the MySQL database. Note that this package is <strong>not needed by most users</strong> and the data is protected by password/IP combinations. Ask details from admins. Install the package in R with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(microbiome)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Install additional dependencies</span>
<span class="kw">source</span>(<span class="st">"http://www.bioconductor.org/biocLite.R"</span>)

<span class="kw">biocLite</span>(<span class="st">"DBI"</span>)
<span class="kw">biocLite</span>(<span class="st">"RPA"</span>)
<span class="kw">biocLite</span>(<span class="st">"svDialogs"</span>)

<span class="kw">library</span>(devtools) <span class="co"># Load the devtools package</span>
<span class="kw">install_github</span>(<span class="st">"microbiome/HITChipDB"</span>) <span class="co"># Install the package</span>
<span class="co"># Also install RMySQL, multicore and tcltk !</span>
<span class="kw">source</span>(<span class="st">"http://www.bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"RMySQL"</span>) <span class="co"># multicore, tcltk?</span>
<span class="co"># Test installation by loading the microbiome package in R</span>
<span class="kw">library</span>(<span class="st">"HITChipDB"</span>)</code></pre></div>
<p>With HITChip, <a href="http://www.computer.org/csdl/trans/tb/2011/01/ttb2011010217-abs.html">fRPA</a> is the recommended preprocessing method. You can add new metadata fields in the template metadata file in your HITChip data folder and exporting it again to tab-separated .tab format. Some standard, self-explanatory field names include ‘sample’, ‘time’, ‘subject’, ‘group’, ‘gender’, ‘diet’, ‘age’. You can leave these out or include further fields. Import HITChip phylotype-level data in <a href="https://github.com/joey711/phyloseq">phyloseq</a> format (note: the precalculated matrices are calculated with detection = 0):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(HITChipDB)
pseq &lt;-<span class="st"> </span>HITChipDB::<span class="kw">read_hitchip</span>(data.directory, <span class="dt">method =</span> <span class="st">"frpa"</span>)$pseq</code></pre></div>
<p>Get higher taxonomic levels, use (on HITChip we use L1/L2 instead of Phylum/Genus):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq.L2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/aggregate_taxa.html">aggregate_taxa</a></span>(pseq, <span class="dt">level =</span> <span class="st">"L2"</span>)
pseq.L1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/aggregate_taxa.html">aggregate_taxa</a></span>(pseq, <span class="dt">level =</span> <span class="st">"L1"</span>)</code></pre></div>
<p>Importing HITChip probe-level data and taxonomy from HITChip output directory (these are not available in the phyloseq object):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">probedata &lt;-<span class="st"> </span>HITChipDB::<span class="kw">read_hitchip</span>(data.directory, <span class="dt">method =</span> <span class="st">"frpa"</span>)$probedata
taxonomy.full &lt;-<span class="st"> </span>HITChipDB::<span class="kw">read_hitchip</span>(data.directory, <span class="dt">method =</span> <span class="st">"frpa"</span>)$taxonomy.full</code></pre></div>
<p>Convert your own data into phyloseq format as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We need to choose the HITChip data level to be used in the analyses</span>
<span class="co"># In this example use HITChip L2 data (note: this is in absolute scale)</span>
res &lt;-<span class="st"> </span><span class="kw">read_hitchip</span>(<span class="dt">method =</span> <span class="st">"frpa"</span>, <span class="dt">data.dir =</span> data.directory)

<span class="co"># Species-level data matrix</span>
otu &lt;-<span class="st"> </span><span class="kw"><a href="../reference/abundances.html">abundances</a></span>(res$pseq)@.Data 

<span class="co"># Corresponding sample metadata</span>
meta &lt;-<span class="st"> </span>res$meta

<span class="co"># Taxonomy</span>
<span class="co"># First get an experimental function from the microbiome package</span>
f &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"inst/extdata/get_hitchip_taxonomy.R"</span>, <span class="dt">package =</span> <span class="st">"microbiome"</span>)
<span class="kw">source</span>(f)
taxonomy &lt;-<span class="st"> </span><span class="kw">get_hitchip_taxonomy</span>(<span class="st">"HITChip"</span>, <span class="st">"filtered"</span>)
taxonomy &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">as.data.frame</span>(taxonomy[, <span class="kw">c</span>(<span class="st">"L1"</span>, <span class="st">"L2"</span>, <span class="st">"species"</span>)]))
<span class="kw">rownames</span>(taxonomy) &lt;-<span class="st"> </span><span class="kw">as.vector</span>(taxonomy[, <span class="st">"species"</span>])

<span class="co"># Merging data matrices into phyloseq format:</span>
pseq &lt;-<span class="st"> </span>HITChipDB::<span class="kw">hitchip2physeq</span>(<span class="kw">t</span>(otu), meta, taxonomy)</code></pre></div>
<div id="refs" class="references">

</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#hitchip-and-other-phylogenetic-microarrays">HITChip and other phylogenetic microarrays</a></li>
      <li><a href="#installing-hitchipdb-package">Installing HITChipDB package</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
