<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Preprocessing of taxonomic profiling data • microbiome</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">microbiome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Preprocessing of taxonomic profiling data</h1>
            
          </div>

    
    
<div class="contents">
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial - Preprocessing}
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}  
-->
<div id="processing-taxonomic-profiling-data" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#processing-taxonomic-profiling-data" class="anchor"> </a></body></html>Processing taxonomic profiling data</h2>
<p>Instructions to manipulate microbiome data sets using tools from the <a href="http://joey711.github.io/phyloseq/">phyloseq package</a> and some extensions from the <a href="https://github.com/microbiome/microbiome">microbiome package</a>, including subsetting, aggregating and filtering.</p>
<p>Load example data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(phyloseq)
<span class="kw">library</span>(microbiome)

<span class="kw">data</span>(atlas1006)   <span class="co"># Load the data</span>
pseq &lt;-<span class="st"> </span>atlas1006 <span class="co"># Rename the data</span></code></pre></div>
<div id="retrieving-data-elements-from-a-phyloseq-object" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#retrieving-data-elements-from-a-phyloseq-object" class="anchor"> </a></body></html>Retrieving data elements from a phyloseq object</h3>
<p>A phyloseq object contains OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and phylogenetic tree (relations between the taxa). Some of these are optional.</p>
<p>Pick metadata:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meta &lt;-<span class="st"> </span><span class="kw">sample_data</span>(pseq)</code></pre></div>
<p>Taxonomy table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">taxonomy &lt;-<span class="st"> </span><span class="kw">tax_table</span>(pseq)</code></pre></div>
<p>Abundances for taxonomic groups (‘OTU table’) as a TaxaxSamples matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Absolute abundances</span>
otu.absolute &lt;-<span class="st"> </span><span class="kw"><a href="../reference/abundances.html">abundances</a></span>(pseq)

<span class="co"># Relative abundances</span>
otu.relative &lt;-<span class="st"> </span><span class="kw"><a href="../reference/abundances.html">abundances</a></span>(pseq, <span class="st">"compositional"</span>)</code></pre></div>
<p>Melt phyloseq data for easier plotting:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">psmelt</span>(pseq)
<span class="kw">kable</span>(<span class="kw">head</span>(df))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">OTU</th>
<th align="left">Sample</th>
<th align="right">Abundance</th>
<th align="right">age</th>
<th align="left">gender</th>
<th align="left">nationality</th>
<th align="left">DNA_extraction_method</th>
<th align="left">project</th>
<th align="right">diversity</th>
<th align="left">bmi_group</th>
<th align="left">subject</th>
<th align="right">time</th>
<th align="left">sample</th>
<th align="left">Phylum</th>
<th align="left">Genus</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">113110</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-448</td>
<td align="right">944002</td>
<td align="right">54</td>
<td align="left">female</td>
<td align="left">CentralEurope</td>
<td align="left">o</td>
<td align="left">18</td>
<td align="right">5.98</td>
<td align="left">lean</td>
<td align="left">448</td>
<td align="right">0</td>
<td align="left">Sample-448</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
<tr class="even">
<td align="left">113015</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-360</td>
<td align="right">902034</td>
<td align="right">45</td>
<td align="left">female</td>
<td align="left">CentralEurope</td>
<td align="left">o</td>
<td align="left">13</td>
<td align="right">5.49</td>
<td align="left">severeobese</td>
<td align="left">360</td>
<td align="right">0</td>
<td align="left">Sample-360</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
<tr class="odd">
<td align="left">112747</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-190</td>
<td align="right">862870</td>
<td align="right">34</td>
<td align="left">female</td>
<td align="left">CentralEurope</td>
<td align="left">r</td>
<td align="left">7</td>
<td align="right">6.06</td>
<td align="left">lean</td>
<td align="left">190</td>
<td align="right">0</td>
<td align="left">Sample-190</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
<tr class="even">
<td align="left">113109</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-743</td>
<td align="right">852350</td>
<td align="right">52</td>
<td align="left">male</td>
<td align="left">US</td>
<td align="left">NA</td>
<td align="left">19</td>
<td align="right">5.21</td>
<td align="left">obese</td>
<td align="left">743</td>
<td align="right">0</td>
<td align="left">Sample-743</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
<tr class="odd">
<td align="left">112944</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-366</td>
<td align="right">851147</td>
<td align="right">52</td>
<td align="left">female</td>
<td align="left">CentralEurope</td>
<td align="left">o</td>
<td align="left">15</td>
<td align="right">5.63</td>
<td align="left">obese</td>
<td align="left">366</td>
<td align="right">0</td>
<td align="left">Sample-366</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
<tr class="even">
<td align="left">113639</td>
<td align="left">Prevotella melaninogenica et rel.</td>
<td align="left">Sample-375</td>
<td align="right">844482</td>
<td align="right">45</td>
<td align="left">female</td>
<td align="left">CentralEurope</td>
<td align="left">o</td>
<td align="left">16</td>
<td align="right">5.64</td>
<td align="left">severeobese</td>
<td align="left">375</td>
<td align="right">0</td>
<td align="left">Sample-375</td>
<td align="left">Bacteroidetes</td>
<td align="left">Prevotella melaninogenica et rel.</td>
</tr>
</tbody>
</table>
</div>
<div id="sample-operations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#sample-operations" class="anchor"> </a></body></html>Sample operations</h3>
<p>Sample names and variables</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">sample_names</span>(pseq))</code></pre></div>
<pre><code>## [1] "Sample-1" "Sample-2" "Sample-3" "Sample-4" "Sample-5" "Sample-6"</code></pre>
<p>Total OTU abundance in each sample</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">sample_sums</span>(pseq))</code></pre></div>
<pre><code>## Sample-1 Sample-2 Sample-3 Sample-4 Sample-5 Sample-6 
##   479428   640574   449884   684997   757697   499535</code></pre>
<p>Abundance of a given species in each sample</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">get_sample</span>(pseq, <span class="st">"Akkermansia"</span>))</code></pre></div>
<pre><code>## Sample-1 Sample-2 Sample-3 Sample-4 Sample-5 Sample-6 
##     1319     2299    29980     3824     2133      864</code></pre>
<p>Filter samples</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f1 &lt;-<span class="st"> </span><span class="kw">filterfun_sample</span>(<span class="kw">topp</span>(<span class="fl">0.1</span>))
taxa &lt;-<span class="st"> </span><span class="kw">genefilter_sample</span>(pseq, f1, <span class="dt">A =</span> <span class="kw">round</span>(<span class="fl">0.5</span> *<span class="st"> </span><span class="kw">nsamples</span>(pseq)))</code></pre></div>
<p>Select samples by specific metadata fields</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq.subset &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(pseq, nationality ==<span class="st"> "US"</span>)</code></pre></div>
<p>Pick samples at the baseline time points only:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/baseline.html">baseline</a></span>(pseq)</code></pre></div>
</div>
<div id="data-transformations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#data-transformations" class="anchor"> </a></body></html>Data transformations</h3>
<p>The microbiome package provides a wrapper for standard sample/OTU transforms. For arbitrary transforms, use the transform_sample_counts function in the phyloseq package.</p>
<p>Log10 transform (log(1+x) if the data contains zeroes). Also “Z”, “clr”, and “hellinger” are available as common transforms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq.log &lt;-<span class="st"> </span><span class="kw"><a href="../reference/transform.html">transform</a></span>(pseq, <span class="st">"log10"</span>)</code></pre></div>
<p>Relative abundances (the input data needs to be in absolute scale, not logarithmic!):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq1 &lt;-<span class="st"> </span>microbiome::<span class="kw"><a href="../reference/transform.html">transform</a></span>(pseq, <span class="st">"compositional"</span>, <span class="st">"OTU"</span>)
pseq2 &lt;-<span class="st"> </span>phyloseq::<span class="kw">transform_sample_counts</span>(pseq, function(x) x/<span class="kw">sum</span>(x))</code></pre></div>
</div>
<div id="variable-operations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#variable-operations" class="anchor"> </a></body></html>Variable operations</h3>
<p>Sample variable names</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sample_variables</span>(pseq)</code></pre></div>
<pre><code>##  [1] "age"                   "gender"               
##  [3] "nationality"           "DNA_extraction_method"
##  [5] "project"               "diversity"            
##  [7] "bmi_group"             "subject"              
##  [9] "time"                  "sample"</code></pre>
<p>Pick values for a given variable</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">get_variable</span>(pseq, <span class="kw">sample_variables</span>(pseq)[<span class="dv">1</span>]))</code></pre></div>
<pre><code>## [1] 28 24 52 22 25 42</code></pre>
<p>Assign new fields to metadata</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Calculate diversity for samples</span>
div &lt;-<span class="st"> </span>microbiome::<span class="kw"><a href="../reference/diversity.html">diversity</a></span>(pseq, <span class="dt">measures =</span> <span class="st">"Shannon"</span>)$Shannon

<span class="co"># Assign this to sample metadata</span>
<span class="kw">sample_data</span>(pseq)$diversity &lt;-<span class="st"> </span>div</code></pre></div>
</div>
<div id="taxa-operations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#taxa-operations" class="anchor"> </a></body></html>Taxa operations</h3>
<p>Number of taxa</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="kw">ntaxa</span>(pseq)</code></pre></div>
<p>Most abundant taxa</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/top_taxa.html">top_taxa</a></span>(pseq, <span class="dt">n =</span> <span class="dv">10</span>)</code></pre></div>
<p>Names</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ranks &lt;-<span class="st"> </span><span class="kw">rank_names</span>(pseq)  <span class="co"># Taxonomic levels</span>
taxa  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/taxa.html">taxa</a></span>(pseq)        <span class="co"># Taxa names at the analysed level</span></code></pre></div>
<p>Subset taxa:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq.bac &lt;-<span class="st"> </span><span class="kw">subset_taxa</span>(pseq, Phylum ==<span class="st"> "Bacteroidetes"</span>)</code></pre></div>
<p>Prune (select) taxa:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># List of Genera in the Bacteroideted Phylum</span>
taxa &lt;-<span class="st"> </span><span class="kw"><a href="../reference/map_levels.html">map_levels</a></span>(<span class="ot">NULL</span>, <span class="st">"Phylum"</span>, <span class="st">"Genus"</span>, pseq)$Bacteroidetes

<span class="co"># With given taxon names</span>
ex2 &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(taxa, pseq)

<span class="co"># Taxa with positive sum across samples</span>
ex3 &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(<span class="kw">taxa_sums</span>(pseq) &gt;<span class="st"> </span><span class="dv">0</span>, pseq)</code></pre></div>
<p>Filter by user-specified function values (here variance):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">filter_taxa</span>(pseq, function(x) <span class="kw">var</span>(x) &gt;<span class="st"> </span><span class="fl">1e-05</span>, <span class="ot">TRUE</span>)</code></pre></div>
<p>List unique phylum-level groups:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">get_taxa_unique</span>(pseq, <span class="st">"Phylum"</span>))</code></pre></div>
<pre><code>## [1] "Actinobacteria"         "Bacilli"               
## [3] "Proteobacteria"         "Verrucomicrobia"       
## [5] "Bacteroidetes"          "Clostridium cluster XV"</code></pre>
<p>Pick the taxa abundances for a given sample:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">samplename &lt;-<span class="st"> </span><span class="kw">sample_names</span>(pseq)[[<span class="dv">1</span>]]

<span class="co"># Two ways to pick abundances for a particular taxon</span>
tax.abundances &lt;-<span class="st"> </span><span class="kw">get_taxa</span>(pseq, samplename)
tax.abundances2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/abundances.html">abundances</a></span>(pseq)[, samplename]</code></pre></div>
</div>
<div id="merging-operations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#merging-operations" class="anchor"> </a></body></html>Merging operations</h3>
<p>Aggregate taxa to higher taxonomic levels. This is particularly useful if the phylogenetic tree is missing. When it is available, see <a href="http://joey711.github.io/phyloseq/merge.html">merge_samples, merge_taxa and tax_glom</a>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/aggregate_taxa.html">aggregate_taxa</a></span>(pseq, <span class="st">"Phylum"</span>) </code></pre></div>
<p>Merging phyloseq objects</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">merge_phyloseq</span>(pseqA, pseqB)</code></pre></div>
</div>
<div id="rarification" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#rarification" class="anchor"> </a></body></html>Rarification</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pseq.rarified &lt;-<span class="st"> </span><span class="kw">rarefy_even_depth</span>(pseq)</code></pre></div>
</div>
<div id="taxonomy" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#taxonomy" class="anchor"> </a></body></html>Taxonomy</h3>
<p>Convert between taxonomic levels (here from Genus (Akkermansia) to Phylum (Verrucomicrobia):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="kw"><a href="../reference/map_levels.html">map_levels</a></span>(<span class="st">"Akkermansia"</span>, <span class="st">"Genus"</span>, <span class="st">"Phylum"</span>, <span class="kw">tax_table</span>(pseq))
<span class="kw">print</span>(m)</code></pre></div>
<pre><code>## [1] "Verrucomicrobia"</code></pre>
</div>
<div id="metadata" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#metadata" class="anchor"> </a></body></html>Metadata</h3>
<p>Visualize frequencies of given factor (sex) levels within the indicated groups (group):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plot_frequencies.html">plot_frequencies</a></span>(<span class="kw">sample_data</span>(pseq), <span class="st">"bmi_group"</span>, <span class="st">"gender"</span>)
<span class="kw">print</span>(res$plot)</code></pre></div>
<p><img src="Preprocessing_files/figure-html/phylogeny-example3-1.png" width="960"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Retrieving the actual data values:</span>
<span class="kw">kable</span>(<span class="kw">head</span>(res$data), <span class="dt">digits =</span> <span class="dv">2</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Groups</th>
<th align="left">Factor</th>
<th align="right">n</th>
<th align="right">pct</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">underweight</td>
<td align="left">female</td>
<td align="right">21</td>
<td align="right">91.30</td>
</tr>
<tr class="even">
<td align="left">underweight</td>
<td align="left">male</td>
<td align="right">2</td>
<td align="right">8.70</td>
</tr>
<tr class="odd">
<td align="left">lean</td>
<td align="left">female</td>
<td align="right">304</td>
<td align="right">61.66</td>
</tr>
<tr class="even">
<td align="left">lean</td>
<td align="left">male</td>
<td align="right">189</td>
<td align="right">38.34</td>
</tr>
<tr class="odd">
<td align="left">overweight</td>
<td align="left">female</td>
<td align="right">102</td>
<td align="right">50.00</td>
</tr>
<tr class="even">
<td align="left">overweight</td>
<td align="left">male</td>
<td align="right">102</td>
<td align="right">50.00</td>
</tr>
</tbody>
</table>
<div id="refs" class="references">

</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#processing-taxonomic-profiling-data">Processing taxonomic profiling data</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
